<a href="https://colab.research.google.com/github/JosenildoJunior/StatPyDataScience/blob/main/Teste_de_Hip%C3%B3tese_e_intervalo_de_confian%C3%A7a_para_vari%C3%A2ncia.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Intervalo de confiança para variância e desvio padrão (1 amostra)**

**Breve Resumo:**
Um intervalo de confiança para a variância de uma população com base em uma amostra única de dados nos fornece uma faixa de valores plausíveis para a variância populacional com um certo nível de confiança.

Quando Utilizar:

Quando desejamos fazer inferências sobre a variabilidade de uma população com base em uma amostra.
Quando queremos especificar um nível de confiança para a estimativa da variância populacional.

Fórmula do Intervalo de Confiança: $
\left( \frac{{(n - 1) \cdot s^2}}{{\chi_{\alpha/2}^2}}, \frac{{(n - 1) \cdot s^2}}{{\chi_{1 - \alpha/2}^2}} \right)$

Onde:

- **n** é o tamanho da amostra.

- **s²** é a variância amostral.

- **χ²/₂²** é o quantil da distribuição qui-quadrado com **n-1** graus de liberdade para o nível de significância **α/2**.

- **χ₁₋α/₂²** é o quantil da distribuição qui-quadrado com **n-1** graus de liberdade para o complemento do nível de significância **1-α/2**.


Vamos observar um exemplo para tentarmos entender melhor

**Exemplo:** Encontre os valores críticos χ²R e χ²L para um intervalo de confiança
de 95% quando o tamanho da amostra é 18.

Primeiramente, vamos começar importando algumas bibliotecas necessárias.

In [ ]:
# Importando o qui quadrado
from scipy.stats import chi2

Agora podemos partir para a parte do cálculo propriamente dito.

In [ ]:
# Dados do problema
n = 18  # Tamanho da amostra

# Graus de liberdade para a distribuição qui-quadrado
df = n - 1

# Nível de confiança
confianca = 0.95

# Intervalo de confiança para a variância
intervalo_variancia = stats.chi2.interval(confianca, df)

# Definindo os valores criticos
crit_superior = intervalo_variancia[1]
crit_inferior = intervalo_variancia[0]

# Exibindo os valores
print(f"χ²R = {crit_superior}, χ²L = {crit_inferior}")

χ²R = 30.19100912163982, χ²L = 7.56418644957757


Dessa forma, conseguimos obter os valores críticos que são bem importantes para definirmos os limites. Vamos observar em outro exemplo.

**Exemplo:** Você seleciona aleatoriamente e pesa as 30 unidades de uma
amostra de um antialérgico. O desvio padrão da amostra é de 1,20 miligramas.
Supondo que os pesos são normalmente distribuídos, construa intervalos de
confiança de 99% para a variância e o desvio padrão da população.

Vamos realizar as importações necessárias.

In [ ]:
# Funções matématicas
import numpy as np
import scipy.stats as stats

# Importando o qui quadrado
from scipy.stats import chi2

In [ ]:
# Dados do problema
n = 30  # Tamanho da amostra
s = 1.20  # Desvio padrão da amostra
confianca = 0.99  # Nível de confiança

# Graus de liberdade para a distribuição qui-quadrado
df = n - 1

Agora que já temos os dados definidos, vamos abordar o restante dos problemas por etapas:

- Vamos definir os valores críticos.
- Calcular os limites para o intervalo.
- Calcular os limites para o desvio padrão amostral.

**Definindo os valores**

In [ ]:
# Intervalo de confiança para a variância
intervalo_variancia = stats.chi2.interval(confianca, df)

# Definindo os valores criticos
crit_superior = intervalo_variancia[1]
crit_inferior = intervalo_variancia[0]

# Exibindo os valores
print(f"χ²R = {crit_superior}, χ²L = {crit_inferior}")

χ²R = 52.335617785933614, χ²L = 13.121148887960413


Para **calcular os limites inferiores e superiores do intervalo de confiança para a variância populacional**, aplicamos a raiz quadrada nos limites inferiores e superiores da variância, pois o desvio padrão é a raiz quadrada da variância. Portanto, as fórmulas são:

$
\text{Limite Inferior} = \frac{(n - 1) \times s^2}{\text{valor crítico superior}}
$

$
\text{Limite Superior} = \frac{(n - 1) \times s^2}{\text{valor crítico inferior}}
$

onde:
- \(n\) é o tamanho da amostra,
- \(s\) é o desvio padrão da amostra, e
- os valores críticos superior e inferior são obtidos a partir da distribuição qui-quadrado com \(n - 1\) graus de liberdade e um determinado nível de confiança.

Para **calcular os limites inferiores e superiores do intervalo de confiança para o desvio padrão populacional**, aplicamos a raiz quadrada nos limites inferiores e superiores da variância, pois o desvio padrão é a raiz quadrada da variância. Portanto, as fórmulas são:

$
\text{Limite Inferior do Desvio Padrão} = \sqrt{\text{Limite Inferior da Variância}}
$

$
\text{Limite Superior do Desvio Padrão} = \sqrt{\text{Limite Superior da Variância}}
$

onde:
- \(n\) é o tamanho da amostra,
- \(s\) é o desvio padrão amostral, e
- os valores críticos superior e inferior são obtidos a partir da distribuição qui-quadrado com \(n - 1\) graus de liberdade e um determinado nível de confiança.


In [ ]:
# Lmites para a variância
limite_inferior_variancia = (n - 1) * s**2 / crit_superior
limite_superior_variancia = (n - 1) * s**2 / crit_inferior

# Limites para o desvio padrão
limite_inferior_desvio = np.sqrt(limite_inferior_variancia)
limite_superior_desvio = np.sqrt(limite_superior_variancia)

print("Intervalo de confiança para a variação populacional:", (limite_inferior_variancia, limite_superior_variancia))
print("Intervalo de confiança para o desvio padrão populacional:", (limite_inferior_desvio, limite_superior_desvio))

Intervalo de confiança para a variação populacional: (0.7979269523636721, 3.1826481321553914)
Intervalo de confiança para o desvio padrão populacional: (0.8932675704197887, 1.783997794885238)


Dessa forma, podemos afirmar que com 99% de confiança, podemos dizer que a variância populacional está entre 0,79 e 3,18, e o desvio padrão populacional entre 0,89 e 1,78 miligramas.

#**Teste de hipótese para a variância (1 amostra)**

**Objetivo:** Comparar a variância da amostra com uma variância hipotética.



**Fórmula:**

- $χ² = (n - 1) * s² / σ²$

**onde:**

- χ² é a estatística do teste, que segue uma distribuição qui-quadrado com n - 1 graus de liberdade.
- n é o tamanho da amostra.
- s² é a variância da amostra.
- σ² é a variância hipotética.

Formulação das Hipóteses

- Hipótese nula (H0): A variância da população é igual à variância hipotética (σ²).
- Hipótese alternativa (H1): A variância da população é diferente da variância hipotética (σ²).

Vamos observar um exemplo para entendermos melhor:

Uma empresa de processamento de laticínios afirma que a
variância da quantidade de gordura no leite integral processado por ela é não
mais que 0,25. Você suspeita que essa afirmação esteja errada e descobre que
uma amostra aleatória de 41 recipientes de leite tem um variância de 0,27.
Para um nível de significância a = 0,05, há evidência suficiente para rejeitar a
afirmação da empresa? Suponha que a população é normalmente
distribuída.

Resposta: Vamos começar definindo as hipóteses:

H0: gordura ≤ 0,25

Ha: gordura > 0,25

Estamos diante de um teste unicaudal à direita.

Agora vamos importar as bibliotecas necessárias.

In [ ]:
# Funções estatísticas
import scipy.stats as stats

Agora podemos partir para o cálculo propriamente dito.

In [ ]:
# Definindo os dados
n = 41 # Tamanho da amostra
s_amostra = 0.27 # Variância amostral
s_hipotetica = 0.25 # Variância afirmada pela empresa
alpha = 0.05 # Nível de significância
df = n - 1 # Graus de liberdade

# Calculando o valor do teste
valor_teste = (n - 1) * s_amostra  / s_hipotetica

# Exibindo o valor
print(valor_teste)

43.2


Agora vamos calcular o p-valor para fazermos as nossas comparações.

In [ ]:
# Calculando o p-valor
p_valor = 1 - stats.chi2.cdf(valor_teste, df)

# Exibindo o p-valor
print("P-valor:", p_valor)

# Conclusões
if p_valor < alpha:
    print("Rejeitar H0: Há evidências suficientes para concluir que a gordura no leite é maior que 0.25")
else:
    print("Falha em rejeitar H0: Não existem evidências suficientes para concluir que a gordura no leite é maior que 0.25")

P-valor: 0.33622313346405086
Falha em rejeitar H0: Não existem evidências suficientes para concluir que a gordura no leite é maior que 0.25


De acordo com o nosso teste de hipótese e com um nível de significância de 0.05, falhamos em rejeitar a hipótese nula. Portanto, podemos concluir que não há evidências suficientes para afirmar que a quantidade de gordura no leite é maior que 0,25.

Vamos observar outro exemplo:

Um fabricante de artigos esportivos afirma que a variância da força de uma certa linha de pesca é de 15,9. Uma amostra aleatória de 15 rolos de linha tem uma variância de 21,8. Para um nível de significância de α = 0,05, há evidência suficiente para rejeitar a afirmação do fabricante? Suponha que a população é normalmente distribuída.

Vamos começar definindo as hipóteses:

H0: força da linha = 15,9

Ha: força da linha ≠ 15,9

Estamos diante de um teste bicaudal.

É importante lembrarmos que quando utilizamos o teste qui-quadrado é muito importante que trabalhemos com uma distribuição normal, ou pelo menos 40 elementos na amostra. Nesse caso, o problema já nos informa que temos uma população normalmente distribuída, então podemos seguir adiante.

Importando a biblioteca necessária.

In [ ]:
# Funções estatísticas
import scipy.stats as stats

Partindo para os cálculos.

In [ ]:
# Definindo os valores
n = 15 # Tamanho da amostra
s_amostra = 21.8 # Variância amostral
s_hipotetica = 15.9 # Variância afirmada
alpha = 0.05 # Nível de significância
df = n - 1 # Graus de liberdade

# Calculando o valor do teste
valor_teste = (n - 1) * s_amostra / s_hipotetica

# Exibindo o valor
print(valor_teste)

19.19496855345912


Agora vamos calcular o p-valor.

In [ ]:
# Calculando o p-valor
p_valor = 2 * (1 - stats.chi2.cdf(valor_teste, df))

# Exibindo o p-valor
print("P-valor:", p_valor)

# Conclusões
if p_valor < alpha:
    print("Rejeitar H0: Há evidências suficientes para concluir que a variância da força da linha é diferente de 15,9.")
else:
    print("Falha em rejeitar H0: Não existem evidências suficientes para concluir que a variância da força da linha é diferente de 15,9.")

P-valor: 0.3152647436788134
Falha em rejeitar H0: Não existem evidências suficientes para concluir que a força da linha é diferente de 15,9.


Segundo nosso teste de hipótese com um nível de significância de 0,05, falhamos em rejeitar a hipótese nula. Portanto, não há evidências suficientes para rejeitar a afirmação de que a variância da força da linha de pesca é igual a 15,9.

#**Teste de hipótese para a variância (2 amostras)**

A fórmula do teste F é utilizada principalmente para comparar a variabilidade entre duas amostras. Ela é comumente empregada em testes de hipóteses para verificar se as variâncias de duas populações são estatisticamente iguais ou diferentes.

Dito isso essa é a fórmula do teste F:

- $F = s1^2 / s2^2$

Onde:

- $s1^2$ é a variância da primeira amostra.
- $s2^2$ é a variância da segunda amostra.

Necessariamente, s1 é maior que s2; então, será dessa forma que identificaremos qual é a primeira amostra e qual é a segunda.

Vamos observar um exemplo:


Você quer comprar ações em uma empresa e está decidindo entre duas
ações diferentes. Como o risco de uma ação pode estar associado ao desvio
padrão dos preços de fechamento diários, você seleciona aleatoriamente
amostras dos preços de fechamento diários para cada ação e obtém os
resultados mostrados na tabela abaixo. Com a = 0,05, você pode concluir que
uma das duas ações é um investimento mais arriscado? Suponha que os
preços de fechamento das ações são normalmente distribuídos.

A ação A apresenta os seguintes dados
- n2 = 30
- s2 = 3,5

A ação B apresenta os seguintes dados
- n1 = 31
- s1 = 5,7


Resposta: Vamos começar definindo nossas hipóteses.

H0: risco da ação A = risco da ação B

Ha: risco da ação A ≠ risco da ação B

Estamos diante de um teste bicaudal.

Como a ação B apresentou uma variância maior, ela será nossa primeira amostra, e a ação A será nossa segunda.

Agora vamos prosseguir com a importação das bibliotecas.

In [ ]:
# Funções matématicas
import numpy as np

# Teste f
import scipy.stats as stats

Agora vamos para o cálculo propriamente dito.

In [ ]:
# Definindo os valores
# Ação A
n2 = 30 # Tamanho da amostra
s2 = 3.5 ** 2 # Variância
df2 = n2 - 1 # Graus de liberdade

# Ação B
n1 = 31 # Tamanho da amostra
s1 = 5.7 ** 2 # Variância
df1 = n1 - 1 # Graus de liberdade

# Calculando o F
F = s1 / s2

# Exibindo o valor
print(F)

2.652244897959184


Agora vamos calcular o p-valor.

In [ ]:
# Definindo o alpha
alpha = 0.05 / 2 # Realizando a divisão por se tratar de um teste bicaudal

# Calculando o p-valor
p_valor = 1 - stats.f.cdf(F, df1, df2)

print("O p-valor é:", p_valor)

# Conclusões
if p_valor < alpha:
    print("Rejeitar H0: Há evidências suficientes para concluir que uma ação é mais arriscada que a outra.")
else:
    print("Falha em rejeitar H0: Não existem evidências suficientes para concluir que uma ação é mais arriscada que a outra.")

O p-valor é: 0.005108622913231997
Rejeitar H0: Há evidências suficientes para concluir que uma ação é mais arriscada que a outra.


Após a execução do nosso teste de hipóteses, rejeitamos a hipótese nula. Logo, podemos concluir que existem evidências para afirmar que uma ação é mais arriscada que a outra.

Para casos como esses é possivel utilizarmos uma calculadora online, onde basta passarmos o nosso F, nosso denominador e o nosso numerador além do nivel de significancia e ela já vai nos retornar o nosso p-valor, esse é o link da ferramenta:

- [Calculadora de Distribuição F](https://www.socscistatistics.com/pvalues/fdistribution.aspx)


# **Teste de hipótese - categóricos ou proporção com mais de 2 categorias (1 amostra)**

O teste qui-quadrado é utilizado para testar se uma distribuição de frequência
observada (valores amostrais) se ajusta a uma distribuição esperada (valores
fixos que queremos comparar). Porém para aplicação deste teste algumas condições devem ser respeitadas tais condições são:

- As frequências observadas devem ser obtidas de uma amostra aleatória.
- Cada frequência esperada deve ser maior ou igual a 5.

Se a situação respeitar essas condições podemos aplicar a seguinte formula:

- x^2 = $\frac{(O - E)^2}{E}$

Onde:

- x: é o valor observado na célula da tabela de contingência.
- O: é o valor observado na célula da tabela de contingência.
- E: é o valor esperado na célula da tabela de contingência.

Lembrando que temos k – 1 graus de liberdade, sendo k o número de categorias.



Vamos observar um exemplo para tentarmos entender melhor.


Uma associação de comércio varejista afirma que os meios de preparação de
imposto são distribuídos conforme a tabela abaixo. Uma consultoria de
impostos seleciona aleatoriamente 300 adultos e pergunta como eles
preparam seus impostos. Os resultados encontram-se na tabela abaixo. Para a
= 0,01, teste a afirmação da associação.

**Distribuição esperada para os meios**

Contador = 24%

Á mão = 20%

Programa de computador = 35%

Amigo/familiar = 6%

Consultoria de impostos 15%

**Resultados da pesquisa (n = 300)**

Contador = 61

Á mão = 42

Programa de computador = 112

Amigo/familiar = 29

Consultoria de impostos = 56

Dito isso vamos definir nossas hipóteses:

H0: A distribuição esperada dos métodos de preparação de impostos é: 24%
por contador (72 do total de 300), 20% (60 do total de 300) à mão, 35% (105 do
total de 300) com programa de computador, 6% (18 do total de 300) por
amigo ou familiar e 15% (45 do total de 300) com consultoria de impostos.

Ha: A distribuição dos métodos difere da distribuição esperada.

Como existem 5 categorias, nosso grau de liberdade será 4, pois temos que: g.l = 5 - 1, logo g.l = 4

Agora podemos partir para a estruturação dos dados e cálculo do qui-quadrado.

In [ ]:
# Definindo os dados em listas
esperado = [72, 60, 105, 18, 45]
observado = [61, 42, 112, 29, 56]

# Função para calcular o teste qui-quadrado para cada categoria
def qui_quadrado(esperado, observado):
    return (observado - esperado) ** 2 / esperado

# Calculando o qui-quadrado para cada categoria
resultados = [qui_quadrado(esperado[i], observado[i]) for i in range(len(esperado))]

# Realizando a soma dos resultados
x = sum(resultados)

# Graus de liberdade
df = 5 - 1

# Nível de significância
alpha = 0.01

# Exibindo o valor
print(x)

16.958333333333336


Agora que temos o nosso qui-quadrado score, vamos buscar o p-valor.

In [ ]:
# Importando a biblioteca
from scipy.stats import chi2

# Calculando o p-valor
p_valor = 1 - chi2.cdf(x, df)

# Conclusões
if p_valor < alpha:
    print("Rejeitar H0: Há evidências suficientes para concluir que a distribuição dos métodos difere da distribuição esperada.")
else:
    print("Falha em rejeitar H0: Não existem evidências suficientes para concluir que a distribuição dos métodos difere da distribuição esperada.")

Rejeitar H0: Há evidências suficientes para concluir que a distribuição dos métodos difere da distribuição esperada.


Com base no nosso teste de hipótese a um nível de significância de 0,01, rejeitamos a hipótese nula. Logo, podemos concluir que existem evidências suficientes de que a distribuição dos métodos difere da distribuição esperada, portanto, são diferentes.

Após calcularmos o qui-quadrado, existe uma calculadora na qual basta passarmos o mesmo e os graus de liberdade, junto com o alpha, que ela já retornará o p-valor. Vou deixar o link da ferramenta caso alguém tenha interesse:

- [Calculadora de Distribuição Qui-Quadrado](https://www.socscistatistics.com/pvalues/chidistribution.aspx)

# **Teste de hipótese - categóricos ou proporção com mais de 2 categorias (2 amostras)**

Nesse caso, também utilizaremos o teste qui-quadrado, porém existem algumas diferenças, como a aplicação da fórmula.

A estatística de teste qui-quadrado é calculada pela seguinte fórmula:
- $\chi^2 = \sum \frac{(O_{ij} - E_{ij})^2}{E_{ij}}$

onde:
- $\chi^2$ é a estatística de teste qui-quadrado,
- $O$ são as frequências observadas,
- $E$ são as frequências esperadas.

Dito isso, vamos observar um Exercício


Exercício:
Um estudo foi realizado para investigar a preferência de bebidas entre dois grupos de pessoas (grupo A e grupo B). Os participantes foram questionados sobre sua preferência entre quatro tipos de bebidas: água, suco de laranja, refrigerante e chá. Os resultados foram os seguintes:

- Grupo A (100 pessoas): 20 preferem água, 30 preferem suco de laranja, 40 preferem refrigerante, 10 preferem chá.
- Grupo B (120 pessoas): 30 preferem água, 25 preferem suco de laranja, 45 preferem refrigerante, 20 preferem chá.

É possivel determinar se há uma associação significativa entre a preferência de bebidas e os grupos A e B

Resposta: Nesse caso, estamos diante de variáveis independentes, logo, podemos aplicar o nosso teste.

Vamos criar as nossas hipóteses:

H0: A = B -> não há diferença significativa na preferência de bebidas entre os grupos A e B

Ha: A ≠ B -> há diferença significativa na preferência de bebidas entre os grupos A e B

In [ ]:
# Importando a biblioteca
from scipy.stats import chi2_contingency

# Dados observados
grupo_A = [20, 30, 40, 10]
grupo_B = [30, 25, 45, 20]

# Criando uma tabela de contingência com os dados observados
tabela_contingencia = [grupo_A, grupo_B]

# Executando o teste qui-quadrado de independência
chi2, p_valor, _, _ = chi2_contingency(tabela_contingencia)

# Imprimindo o resultado
print("Valor qui-quadrado:", chi2)
print("p-valor:", p_valor)

Valor qui-quadrado: 4.299346405228759
p-valor: 0.23090173858289798


Dessa forma, conseguimos obter o p-valor e o qui-quadrado. Agora podemos tirar nossas conclusões.

In [ ]:
# Definindo o alpha
alpha = 0.05

# Conclusões
if p_valor < alpha:
    print("Rejeitar H0: NÂO existe evidências suficientes para concluir que existe uma associação significativa entre as preferência de bebidas e os grupos A e B.")
else:
    print("Falha em rejeitar H0: Existem evidências suficientes para concluir que existe uma associação significativa entre as preferência de bebidas e os grupos A e B.")

Falha em rejeitar H0: Existem evidências suficientes para concluir que existe uma associação significativa entre as preferência de bebidas e os grupos A e B.


Dado o nosso teste de hipótese a um nível de significância de 0.05, falhamos em rejeitar a hipótese nula. Logo, podemos concluir que existem evidências de que não existe uma associação significativa entre a preferência de bebidas e os grupos A e B.

Uma outra opção seria calcularmos o teste e utilizarmos a calculadora online para calcularmos o p-valor. Vou deixar o link da mesma:

- [Calculadora de Distribuição Qui-Quadrado](https://www.socscistatistics.com/pvalues/chidistribution.aspx)

# **ANOVA um fator**

Breve Resumo:

A análise de variância (ANOVA) é uma técnica estatística utilizada para comparar as médias de três ou mais grupos independentes. Ela avalia se existe uma diferença significativa entre as médias dos grupos e é comumente utilizada quando se tem um fator categórico com mais de duas categorias.

Quando Utilizar:

- Quando você deseja comparar as médias de três ou mais grupos independentes para determinar se há uma diferença estatisticamente significativa entre eles.
- Ao investigar o efeito de um único fator categórico com mais de duas categorias sobre uma variável contínua.
- Quando você tem dados que atendem aos pressupostos da ANOVA, como normalidade, homogeneidade das variâncias e independência entre as observações.

Requisitos para aplicação:

- Independência das observações: As observações em cada grupo devem ser independentes umas das outras.
- Normalidade dos resíduos: As diferenças entre os valores observados e os valores previstos devem ser normalmente distribuídas em cada grupo.
- Homogeneidade das variâncias: As variâncias dos resíduos devem ser aproximadamente iguais em todos os grupos.

Importante frisarmos que por padrão assumimos as seguintes hipóteses:

- H0: Todas as médias do grupo são iguais.

- Ha: Nem todas as médias do grupo são iguais.

Para um melhor entendimento, vamos observar um exemplo:

Suponha que um pesquisador esteja investigando o efeito de diferentes tipos de adubos (A, B, C) no crescimento de plantas. O pesquisador selecionou aleatoriamente 10 plantas e aplicou um tipo de adubo em cada planta. Após um período de crescimento, ele mediu a altura das plantas em centímetros. Os resultados estão listados abaixo:

- Adubo A: [30, 32, 31, 28, 29, 30, 32, 31, 30, 31]
- Adubo B: [29, 30, 31, 33, 30, 31, 30, 31, 32, 31]
- Adubo C: [27, 28, 30, 29, 28, 30, 29, 28, 29, 30]

Determine se existe uma diferença significativa entre as médias das alturas das plantas para os três tipos de adubos. Use um nível de significância de 0.05.

Vamos definir as hipóteses:

H0: Altura média A = Altura média B = Altura média C  

Ha: Existem diferenças entre as alturas médias

Nesse caso, como temos 3 grupos independentes e queremos saber se existem diferenças significativas entre eles, podemos aplicar o teste ANOVA de um fator. Dito isso, vamos começar com as importações necessárias.

In [ ]:
# Importando o Anova um fator
from scipy.stats import f_oneway

Dito isso, vamos partir para o cálculo.

In [ ]:
# Definindo os grupos
adubo_a = [30, 32, 31, 28, 29, 30, 32, 31, 30, 31]
adubo_b = [29, 30, 31, 33, 30, 31, 30, 31, 32, 31]
adubo_c = [27, 28, 30, 29, 28, 30, 29, 28, 29, 30]

# Executando ANOVA de um fator
f_statistic, p_value = f_oneway(adubo_a, adubo_b, adubo_c)

print("Estatística F:", f_statistic)
print("Valor p:", p_value)

if p_value < 0.05:
    print("Rejeitar H0: Há diferença significativa entre as médias.")
else:
    print("Falha em rejeitar H0: Não há diferença significativa entre as médias.")

Estatística F: 8.494382022471909
Valor p: 0.0013749821478992777
Rejeitar H0: Há diferença significativa entre as médias.


Dessa forma, ao aplicarmos o teste e considerando nosso nível de significância de 0,05, rejeitamos a hipótese nula. Logo, podemos concluir que existem evidências de que as médias das alturas das plantas são diferentes.

## **Teste de Tukey**






O teste de Tukey é uma análise de comparação múltipla usada após o teste ANOVA. Ele nos permite determinar quais grupos têm médias significativamente diferentes entre si.

Passos para realizar o teste de Tukey:

- Independência das observações: Mesmo pressuposto que o teste ANOVA.
- Normalidade dos resíduos: Mesmo pressuposto que o teste ANOVA.
- Homogeneidade das variâncias: Mesmo pressuposto que o teste ANOVA.

Realizar o teste de Tukey:

- Se os pressupostos forem atendidos, o teste de Tukey pode ser aplicado diretamente para comparar todas as médias dos grupos.

- Se os pressupostos não forem atendidos (por exemplo, desigualdade de variâncias), considerar o teste de Games-Howell como uma alternativa.

Vamos observar a implementação do mesmo, já que a ANOVA já foi aplicada.

In [ ]:
# Importando o teste
from statsmodels.stats.multicomp import pairwise_tukeyhsd

# Juntando os dados em um único DataFrame
data = pd.DataFrame({
    'altura': adubo_a + adubo_b + adubo_c,
    'adubo': ['A'] * 10 + ['B'] * 10 + ['C'] * 10 # Como os primeiros 10 registros são do grupo A estou multiplicando o A 10 vezes para ele aparecer em todos os seus elementos
                                                  # O mesmo esta sendo aplicado para as outras duas categórias
})

# Aplicando o teste de Tukey
tukey_results = pairwise_tukeyhsd(data['altura'], data['adubo'])

# Exibindo o resultado
print(tukey_results)

Multiple Comparison of Means - Tukey HSD, FWER=0.05 
group1 group2 meandiff p-adj   lower   upper  reject
----------------------------------------------------
     A      B      0.4 0.7189 -0.8732  1.6732  False
     A      C     -1.6 0.0116 -2.8732 -0.3268   True
     B      C     -2.0 0.0016 -3.2732 -0.7268   True
----------------------------------------------------


Dessa forma, podemos observar as comparações entre cada grupo. Dito isso, vamos ver o que conseguimos extrair desse resultado:



**Comparação entre os grupos A e B:**

- Podemos observar que a diferença da média é de 0,4.
- O p-valor ajustado é de 0,71, o que é maior que o nosso alfa de 0,05.
- O intervalo de confiança para a diferença da média varia de -0,87 a 1,67.
- Como o p-valor ajustado é maior que o nosso alfa, falhamos em rejeitar a hipótese nula. Logo, não existe diferença significativa entre os grupos A e B.


**Comparação entre os grupos A e C:**

- Podemos observar que a diferença da média é de -1,6.
- O p-valor ajustado é de 0,01, o que é menor que o nosso alfa de 0,05.
- O intervalo de confiança para a diferença da média varia de -2,87 a -0,32.
- Como o p-valor ajustado é menor que o nosso alfa, rejeitamos a hipótese nula, logo podemos concluir que existe diferença significativa entre os grupos A e C.


**Comparação entre os grupos B e C:**

- Podemos observar que a diferença da média é de -2,0.
- O p-valor ajustado é de 0,001, o que é menor que o nosso alfa de 0,05.
- O intervalo de confiança para a diferença da média varia de -3,27 a -0,72.
- Como o p-valor ajustado é menor que o nosso alfa, rejeitamos a hipótese nula, logo podemos concluir que existe diferença significativa entre os grupos B e C.

Dessa forma, podemos concluir, com um nível de significância de 0,05, que existem evidências da existência de diferenças estatísticas entre os grupos A:C e B:C.

## **Teste de Games-Howell**



O teste de Games-Howell é uma alternativa ao teste de Tukey quando os pressupostos do ANOVA não são atendidos, especialmente quando há desigualdade de variâncias entre os grupos.

Passos para realizar o teste de Games-Howell:

- Independência das observações: Mesmo pressuposto que o teste ANOVA.

- Normalidade dos resíduos: Mesmo pressuposto que o teste ANOVA.

- Homogeneidade das variâncias: Não é necessário assumir homogeneidade das variâncias entre os grupos.

Realizar o teste de Games-Howell para comparações múltiplas entre os grupos.

Digamos que os dados que temos não passaram em todos os pressupostos para a aplicação de um teste ANOVA, nesse caso vamos aplicar o teste de Games-Howell dito isso vamos realizar algumas importações necessarias

In [ ]:
# Realizando a instalação
!pip install pingouin

# Funções matématicas
import numpy as np

# Manipulação de dados
import pandas as pd

# Importando a biblioteca pingouin
import pingouin as pg

Com todas as bibliotecas importadas podemos passar para a parte do calculo propriamente dito

In [ ]:
# Dados fornecidos
grupo_A = [30, 32, 31, 28, 29, 30, 32, 31, 30, 31]
grupo_B = [29, 30, 31, 33, 30, 31, 30, 31, 32, 31]
grupo_C = [27, 28, 30, 29, 28, 30, 29, 28, 29, 30]

    # Criando um DataFrame Pandas
data = pd.DataFrame({
        'Altura': np.concatenate([grupo_A, grupo_B, grupo_C]),
'Adubo': ['A'] * len(grupo_A) + ['B'] * len(grupo_B) + ['C'] * len(grupo_C)
})

# Realizando o teste de Games-Howell
games_howell_results = pg.pairwise_gameshowell(data, dv='Altura', between='Adubo')

# Exibindo o resultado
print(games_howell_results)

   A  B  mean(A)  mean(B)  diff        se         T         df      pval  \
0  A  B     30.4     30.8  -0.4  0.537484 -0.744208  17.793636  0.740922   
1  A  C     30.4     28.8   1.6  0.516398  3.098387  17.307692  0.016777   
2  B  C     30.8     28.8   2.0  0.485341  4.120817  17.841214  0.001795   

     hedges  
0 -0.318757  
1  1.327092  
2  1.765017  


Comparação entre os grupos A e B:

- A média do grupo A é 30.4 e a média do grupo B é 30.8.
- A diferença entre as médias é -0.4.
- O erro padrão da diferença é 0.537.
- A estatística de teste é -0.744.
- Os graus de liberdade são 17.79.
- O valor-p é 0.741, o que indica que a diferença não é estatisticamente significativa.
- O tamanho do efeito (Hedges) é -0.319.

Comparação entre os grupos A e C:

- A média do grupo A é 30.4 e a média do grupo C é 28.8.
- A diferença entre as médias é 1.6.
- O erro padrão da diferença é 0.516.
- A estatística de teste é 3.098.
- Os graus de liberdade são 17.31.
- O valor-p é 0.017, o que indica que a diferença é estatisticamente significativa.
- O tamanho do efeito (Hedges) é 1.327.

Comparação entre os grupos B e C:

- A média do grupo B é 30.8 e a média do grupo C é 28.8.
- A diferença entre as médias é 2.0.
- O erro padrão da diferença é 0.485.
- A estatística de teste é 4.121.
- Os graus de liberdade são 17.84.
- O valor-p é 0.002, o que indica que a diferença é estatisticamente significativa.
- O tamanho do efeito (Hedges) é 1.765.

Utilizamos os mesmos dados e obtivemos um resultado muito semelhante. Novamente, os únicos grupos que apresentam diferenças estatisticamente significativas são os grupos A:C e B:C.

# **Exercícios**

## **Questão 1**

Uma empresa de e-commerce deseja analisar a variabilidade no tempo de entrega de seus produtos para diferentes regiões. Eles coletaram uma amostra de 30 tempos de entrega para a Região A e desejam calcular um intervalo de confiança para a variância populacional do tempo de entrega nessa região.

**Tempo de entrega (em dias) para a Região A - [3, 4, 5, 6, 4, 5, 5, 7, 6, 4, 5, 6, 7, 4, 5, 6, 5, 4, 6, 7, 5, 4, 6, 5, 7, 6, 5, 4, 5, 6]**

Calcule um intervalo de confiança de 95% para a variância do tempo de entrega na Região A. O que esse intervalo representa em termos de variabilidade nos tempos de entrega?

Vamos utilizar o qui-quadrado. Dito isso, vamos começar realizando as importações necessárias.

In [ ]:
# Importando o qui quadrado
from scipy.stats import chi2

# Funções matématicas
import numpy as np

Agora vamos definir os dados e partir para o cálculo.

In [ ]:
# Definindo os dados
regiao_a =  [3, 4, 5, 6, 4, 5, 5, 7, 6, 4, 5, 6, 7, 4, 5, 6, 5, 4, 6, 7, 5, 4, 6, 5, 7, 6, 5, 4, 5, 6]

# Tamanho da amostra
n = len(regiao_a)

# Definindo os graus de liberdade
df = n - 1

# Definindo a variância amostral
s = np.var(regiao_a)

# Definindo o nível de confiança
confianca = 0.95

# Intervalo de confiança para a variância
intervalo_variancia = chi2.interval(confianca, df)

# Definindo os valores criticos
crit_superior = intervalo_variancia[1]
crit_inferior = intervalo_variancia[0]

# Exibindo os valores
print(f"χ²R = {crit_superior}, χ²L = {crit_inferior}")

χ²R = 45.72228580417452, χ²L = 16.047071695364906


Agora que conseguimos os nossos valores críticos, basta seguirmos para a definição dos nossos limites.

In [ ]:
# Lmites para a variância
limite_inferior_variancia = (n - 1) * s**2 / crit_superior
limite_superior_variancia = (n - 1) * s**2 / crit_inferior

print("Intervalo de confiança para a variação populacional:", (limite_inferior_variancia, limite_superior_variancia))

Intervalo de confiança para a variação populacional: (0.7846088454586377, 2.235554907310923)


Dessa forma, podemos afirmar que, com 95% de confiança, podemos dizer que a variância populacional está entre 0,78 e 2,23.

## **Questão 2**

Uma empresa de manufatura está interessada em verificar se a variância dos diâmetros de um determinado componente atende às especificações de qualidade. A variância máxima permitida é de 0,025. A empresa coletou uma amostra de e deseja realizar um teste de hipótese para determinar se a variância dos diâmetros difere significativamente da especificação.

Diâmetros dos componentes - [4.2, 4.3, 4.1, 4.4, 4.2, 4.4, 4.3, 4.2, 4.5, 4.4, 4.2, 4.3, 4.1, 4.4, 4.5, 4.3, 4.3, 4.4, 4.4, 4.2, 4.4, 4.3, 4.0, 4.1, 3.9, 4.4, 4.3, 4.6, 4.2, 4.0, 4.4, 4.0, 4.3]

A variância dos diâmetros dos componentes difere significativamente da especificação?

Resposta: Sabendo disso, vamos começar definindo as nossas hipóteses:

H0: variância <= especificação

Ha: variância > especificação

Estamos diante de um teste unicaudal à direita.

Nesse caso, vamos aplicar o teste qui-quadrado. Dito isso, vamos seguir com as importações.

In [ ]:
# Funções estatísticas
import scipy.stats as stats

Agora vamos partir para os cálculos.

In [ ]:
# Definindo os valores
amostra = [4.2, 4.3, 4.1, 4.4, 4.2, 4.4, 4.3, 4.2, 4.5, 4.4, 4.2, 4.3, 4.1, 4.4, 4.5, 4.3, 4.3, 4.4, 4.4, 4.2, 4.4, 4.3, 4.0, 4.1, 3.9, 4.4, 4.3, 4.6, 4.2, 4.0, 4.4, 4.0, 4.3]

# Tamanho da amostra
n = len(amostra)

# Variância amostral
s_amostra = np.var(amostra)

# Variância afirmada
s_hipotetica = 0.025

# Nível de significância
alpha = 0.05

# Graus de liberdade
df = n - 1

# Calculando o valor do teste
valor_teste = (n - 1) * s_amostra / s_hipotetica

# Exibindo o valor
print(valor_teste)

32.7933884297521


In [ ]:
# Outra maneira de conseguir a variância

"""import statistics

# Lista de valores
lista = [1, 2, 3, 4, 5]

# Calculando a variância usando statistics
variancia = statistics.variance(amostra)

print("Variância:", variancia)"""

Variância: 0.02642045454545457


Agora que temos o valor do teste, vamos calcular o p-valor.

In [ ]:
# Calculando o p-valor
p_valor = 1 - stats.chi2.cdf(valor_teste, df)

# Exibindo o p-valor
print("P-valor:", p_valor)

# Conclusões
if p_valor < alpha:
    print("Rejeitar H0: Há evidências suficientes para concluir que a variância é maior que 0,025.")
else:
    print("Falha em rejeitar H0: Não existem evidências suficientes para concluir que a variância é maior que 0,025.")

P-valor: 0.4279283532700553
Falha em rejeitar H0: Não existem evidências suficientes para concluir que a variância é maior que 0,025.


Segundo nosso teste de hipótese e levando em consideração um nível de significância de 0,05, falhamos em rejeitar a hipótese nula, logo não temos evidências suficientes para concluir que a variância é maior do que o valor especificado.

#**Questão 3**

Uma empresa de alimentos lançou três diferentes embalagens para um de seus produtos populares. Eles coletaram dados de vendas durante um mês em diferentes regiões para avaliar se existe diferença significativa nas vendas entre as três embalagens.

Vendas do produto para as três embalagens em diferentes regiões.

| Região | Embalagem 1 | Embalagem 2 | Embalagem 3 |
| --- | --- | --- | --- |
| Região A | 120 | 130 | 110 |
| Região B | 150 | 140 | 160 |
| Região C | 100 | 110 | 105 |
| Região D | 135 | 125 | 130 |

Determine se existe uma diferença significativa nas vendas entre as três embalagens. Caso haja diferença, identifique quais embalagens têm médias de vendas significativamente diferentes e qual é o insight prático que pode ser obtido a partir desses resultados.

Reposta: Como temos mais de 2 amostras podemos utilizar o ANOVA, mas antes, vamos ver se esses dados seguem os requisitos para aplicarmos o teste:

- Amostras aleatorias

- Grupos independentes

- A variável dependente é continua

- Amostras tem pelo menos n = 15 cada

- Variâncias populacionais em cada grupo é igual

Ele segue quase todos os requisitos, basta conferirmos as variancias nesse caso utilizaremos o **Barlett** ele assume que a hipótese nula para o teste é que as variâncias são iguais para todas as amostras. Já a hipótese alternativa assume que as variâncias não são iguais para um par ou mais Esse teste pode ser usado se tiver dados normalmente distribuidos ou +20 amostras

Dito isso vamos realizar as importações necessarias

In [6]:
# Manipulação de dados
import pandas as pd

# Importando o Barlett
from scipy.stats import bartlett

Agora vamos criar um DF utilizando os dados da questão

In [8]:
# Dados
dados = {
    "Região": ["Região A", "Região B", "Região C", "Região D"],
    "Embalagem 1": [120, 150, 100, 135],
    "Embalagem 2": [130, 140, 110, 125],
    "Embalagem 3": [110, 160, 105, 130]
}

# Criar DataFrame
df = pd.DataFrame(dados)

# Exibir DataFrame
df

Região  Embalagem 1  Embalagem 2  Embalagem 3
0  Região A          120          130          110
1  Região B          150          140          160
2  Região C          100          110          105
3  Região D          135          125          130

Antes de prosseguirmos vamos realizar uma rapida analise nas médias

In [20]:
# Resumos estatisticos
df.describe()

Embalagem 1  Embalagem 2  Embalagem 3
count     4.000000         4.00     4.000000
mean    126.250000       126.25   126.250000
std      21.360009        12.50    24.958299
min     100.000000       110.00   105.000000
25%     115.000000       121.25   108.750000
50%     127.500000       127.50   120.000000
75%     138.750000       132.50   137.500000
max     150.000000       140.00   160.000000

Dessa forma podemos observar que as médias das 3 embalagens são iguas, todas elas são 126,25. Isso já pode ser um indicativo do que vamos achar quando concluirmos o nosso teste, provavelmente o resultado será que não existe diferença entre as médias, mas só para confirmarmos vamos continuar com o raciocinio

**Barlett**

Agora vamos aplicar o Barlett propriamente dito

In [21]:
# Teste de Barlett
statistic, p_valor = bartlett(df['Embalagem 1'], df['Embalagem 2'], df['Embalagem 3'])

# Resultados
print("Estatística de teste de Barlett:", statistic)
print("Valor-p:", p_valor)

# Verificação da hipótese nula
alpha = 0.05
if p_value < alpha:
    print("Rejeitar H0: Não existe evidências suficientes para concluirmos que as variâncias são iguais.")
else:
    print("Falha em rejeitar H0: Há evidências suficientes para concluirmos que as variâncias são iguais.")

Estatística de teste de Barlett: 1.182973421092691
Valor-p: 0.5535037728277683
Falha em rejeitar H0: Há evidências suficientes para concluirmos que as variâncias são iguais.


Já que falhamos em rejeitar H0 podemos concluir com um nivel de significancia de 0,05 de que as variancias são iguais, dito isso vamos aplicar o ANOVA

**ANOVA**

In [22]:
# Importando o Anova um fator
from scipy.stats import f_oneway

In [23]:
# Executando ANOVA de um fator
f_statistic, p_valor = f_oneway(df['Embalagem 1'], df['Embalagem 2'], df['Embalagem 3'])

print("Estatística F:", f_statistic)
print("Valor p:", p_valor)

if p_valor < 0.05:
    print("Rejeitar H0: Há diferença significativa entre as médias.")
else:
    print("Falha em rejeitar H0: Não há diferença significativa entre as médias.")

Estatística F: 0.0
Valor p: 1.0
Falha em rejeitar H0: Não há diferença significativa entre as médias.


O ANOVA confirmou as nossas suspeitas, podemos afirmar a um nivel de significancia de 0,05 que as médias não são estatisticamente diferentes, como não temos diferenças entre as médias o teste de tukey não é necessario, mas mesmo assim vamos ver como seria a sua aplicabilidade

**Tukey**

In [24]:
# Importando o teste
from statsmodels.stats.multicomp import pairwise_tukeyhsd

In [27]:
# Transformando os dados em formato apropriado para o teste de Tukey
stacked_data = df.melt(id_vars=['Região'], value_vars=['Embalagem 1', 'Embalagem 2', 'Embalagem 3'], var_name='Embalagem', value_name='Valor')

# Aplicando o teste de Tukey
tukey_results = pairwise_tukeyhsd(stacked_data['Valor'], stacked_data['Embalagem'])

# Exibindo o resultado
print(tukey_results)

     Multiple Comparison of Means - Tukey HSD, FWER=0.05      
   group1      group2   meandiff p-adj  lower    upper  reject
--------------------------------------------------------------
Embalagem 1 Embalagem 2      0.0   1.0 -40.0634 40.0634  False
Embalagem 1 Embalagem 3      0.0   1.0 -40.0634 40.0634  False
Embalagem 2 Embalagem 3      0.0   1.0 -40.0634 40.0634  False
--------------------------------------------------------------


O Tukey também nos mostra que não existe diferenças entre as médias. Logo podemos concluir definitivamente que não existem diferenças estatisticamente significativas nas médias de vendas das embalagens a um nivel de significancia de 0,05

**Conclusão**:

Como as embalagens não apresentam diferenças significativas entre as médias de vendas é possivel afirmar que o desempenho de vendas das 3 embalagens estão bem proximas, indicando assim uma estabilidade nesse aspecto e abrindo espaço para o foco em outras areas do negocio além de abrir espaço para uma possivel otimização em alguns processos existentes